---
title: "Visualizing and Biclustering scRNA-seq data with CAbiNet"
author:
- name: Clemens Kohl (contributed equally)
  equal_contrib: yes
  affiliation: Max Planck Institute for Molecular Genetics, Berlin, Germany
  email: kohl@molgen.mpg.de
- name: Yan Zhao (contributed equally)
  equal_contrib: yes
  affiliation: Max Planck Institute for Molecular Genetics, Berlin, Germany
  email: zhao@molgen.mpg.de
- name: Martin Vingron
  affiliation: Max Planck Institute for Molecular Genetics, Berlin, Germany
  email: vingron@molgen.mpg.de
package: CAbiNet
output:
  BiocStyle::html_document
abstract: |
  CAbiNet allows fast and robust biclustering of 'marker' genes and cells in single-cell RNA
  sequencing (scRNA-seq) data analysis through Correspondence Analysis. The cells and genes together with the
  biclustering results can be simultaneously displayed in a planar space, the biMAP. CAbiNet can
  facilitate the cell type annotation of scRNA-seq analysis.
vignette: |
  %\VignetteIndexEntry{CAbiNet: Joint visualization of cells and genes based on a gene-cell graph}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options:
  markdown:
    wrap: sentence
---

```{r, echo = FALSE, include = FALSE}
knitr::opts_chunk$set(collapse = TRUE, comment = "#>", results = "hold")
```

# Introduction

CAbiNet allows to jointly visualize cells and genes from scRNA-seq data in a single planar plot, called biMAP. CAbiNet performs biclustering to identify cell groups and their corresponding marker genes, which can be conveniently displayed in the biMAP.

For a more in-depth explanation of CAbiNet, please refer to the pre-print on bioRxiv: https://doi.org/10.1101/2022.12.20.521232

If you are working with `r BiocStyle::Biocpkg("CAbiNet")` please cite:
```
Zhao, Y., Kohl, C., Rosebrock, D., Hu, Q., Hu, Y., Vingron, M.,
CAbiNet: Joint visualization of cells and genes based on a gene-cell graph.
2022, bioRxiv, https://doi.org/10.1101/2022.12.20.521232

```

# Installation

You can install the package from GitHub by running:
```{r git_install, eval=FALSE}
# Until changes are merged into the Bioconductor version, please install the below version of APL.
devtools::install_github("VingronLab/APL", ref = "cabinet")
devtools::install_github("VingronLab/CAbiNet")
```

CAbiNet also has a number of python dependencies, namely umap, umap-learn, numpy, torch and scipy.
They should be installed automatically during the installation of CAbiNet. If there is an issue, you can also install them manually by running the following:

```{r reticulate, eval=FALSE}
library(reticulate)
install_miniconda()
use_condaenv(condaenv = file.path(miniconda_path(),"envs/r-reticulate"),
             required=TRUE)

conda_install(envname = "r-reticulate", packages = "numpy")
conda_install(envname = "r-reticulate", packages = "torch")
conda_install(envname = "r-reticulate", packages = "umap")
conda_install(envname = "r-reticulate", packages = "umap-learn")
conda_install(envname = "r-reticulate", packages = "scipy")
```

If you run into any problem when installing and running our package, please let us know by opening an issue on our GitHub repository https://github.com/VingronLab/CAbiNet/issues. We will help you to solve the problem as best as we can.


# Analyzing data with CAbiNet

CAbiNet is designed to work with a count matrix as input, but is also fully compatible with `r BiocStyle::Biocpkg("SingleCellExperiment")` objects. We will demonstrate here on the Zeisel Brain Dataset obtained from the `r BiocStyle::Biocpkg('scRNAseq')` package how to work with our algorithm and package. This dataset is in a `r BiocStyle::Biocpkg('SingleCellExperiment')` (SCE) object format. If you are working with `r BiocStyle::Rpackage('Seurat')` object, you can still use it as input to `cacomp` function from `r BiocStyle::Biocpkg("APL")`.

In the examples below we will prepare the data and then use CAbiNet to bicluster and visualize the cells and their marker genes. We will further use the biMAP and the biclustering information to quickly and easily annotate the cell types.

## Setup

```{r setup, message=FALSE, warning=FALSE}
library(CAbiNet)
library(APL)
library(scRNAseq)
library(scran)
library(scater)

set.seed(2358)

```

Loading the data:
```{r zeisel}
sce <- ZeiselBrainData()
sce
```

## Data pre-processing

CAbiNet is based on Correspondence Analysis, which is sensitive to outliers in the count matrix. We therefore strongly suggest to pre-process your data before applying CAbiNet. Here we offer some routine pre-processing steps for scRNA-seq data analysis with packages `r BiocStyle::Biocpkg('scran')` and `r BiocStyle::Biocpkg('scater')`. You can modify these steps according to your own demands.
```{r preprocess}

mt_genes <- grepl("^mt-",  rownames(sce), ignore.case = TRUE)

qc_df <- perCellQCMetrics(sce, subsets=list(Mito=mt_genes))

reasons <- perCellQCFilters(qc_df,
                            sum.field = "sum",
                            detected.field = "detected",
                            sub.fields=c("subsets_Mito_percent"))

colData(sce) <- cbind(colData(sce), qc_df)
sce$discard <- reasons$discard

sce <- sce[,!reasons$discard]

cnts<- as.matrix(counts(sce))
genes_detect <- rowSums(cnts > 0) > (ncol(cnts)*0.01)
sce <- sce[genes_detect,]

clust <- quickCluster(sce)
sce <- computeSumFactors(sce, cluster=clust, min.mean=0.1)
sce <- logNormCounts(sce)

dec <- modelGeneVar(sce)
top_genes <- getTopHVGs(dec, prop = 0.8)
sce <- fixedPCA(sce, subset.row=top_genes)
sce <- runUMAP(sce, dimred="PCA")

plotReducedDim(sce, dimred="UMAP", colour_by="level1class")

```

## Dimensionality Reduction in CAbiNet

The dimensionality reduction in CAbiNet is performed with Correspondence Analysis (CA) by the package `r BiocStyle::Biocpkg('APL')`. The most important parameters in the dimensionality reduction step are the number of dimensions (`dims`) you want to keep (similar with the number of principal components in PCA) and the number of genes (`top`) with the highest inertia you want to keep. `top` is set to 5000 by default, in which case the top 5000 genes with highest inertia will be selected before performing CA. If you already picked the most highly variable genes in advance (as we did above), you can keep all genes in the count matrix which will be handed to the function.

The number of genes kept should no be set too low, as otherwise important marker genes are often removed from the data.

```{r CA_dimRed}
# Correspondence Analysis
caobj = cacomp(sce[top_genes,],
               assay = "logcounts",
               dims = 40,
               top = nrow(sce[top_genes,]),
               python = TRUE)
```
The parameter `python = TRUE` indicates that the singular value decomposition in CA is performed with a pytorch implementation. This can drastically speed up the computation, in particular for large datasets. If you did not install all python dependencies for APL, please set it to `FALSE`.

## CAbiNet Biclustering

CAbiNet builds up a shared-nearest-neighbour (SNN) gene-cell graph based on the dimensionality reduced CA space. To build up the graph, you need to specify the number of nearest neighours (NNs) of cells and genes in the function `caclust` with the parameter `k`. As the cell-gene graph is made up of 4 sub-graphs (cell-cell, gene-gene, cell-gene and gene-cell graph), there are also 4 different `k`s that can be tuned. If `k` is set to a single integer, CAbiNet will use the same k for all 4 sub-graphs. If, however, a vector of 4 integers is supplied, it will use them as `k` for the cell-cell, gene-gene, cell-gene and gene-cell graph in that exact order. So, for example `k = 30` is equivalent to `k = (30, 30, 30, 30)`, but `k = c(30, 15, 30, 30)` uses a k half as large as for the other graphs for the gene-gene graph. See also `?caclust`.

The simplest way to run CAbiNet is therefore:
```{r biclustering, eval = FALSE}
# SNN graph & biclustering
cabic <- caclust(obj = caobj,
                 k = c(70, 35, 70, 70),
                 select_genes = FALSE,
                 algorithm = 'leiden',
                 resolution = 1)
```

If a large number of genes have been kept during pre-processing, the biMAP can become too crowded and it can be hard to identify marker genes of interest.
CAbiNet therefore includes a method to remove genes that do not add much information to the biMAP. The paramter `select_genes` indicates whether genes that have no edge to any cell should be removed from the graph:

```{r biclustering_select, eval = FALSE}
# SNN graph & biclustering
cabic <- caclust(obj = caobj,
                 k = c(70, 35, 70, 70),
                 select_genes = TRUE,
                 prune_overlap = FALSE,
                 algorithm = 'leiden',
                 resolution = 1)
```

Furthermore, we offer an optional graph pruning procedure with parameters `prune_overlap` and `overlap` to remove sporadic edges between cells and genes. Setting `prune_overlap` as `TRUE` and `overlap` as a fraction between 0 and 1 allows you to trim out edges between cells and genes that are shared with less than 'overlap'-fraction cell neighours. Note this function only works when `select_genes` is set as `TRUE`.

In many cases we also have a set of marker genes that we would like to use for cell type annotation. In order to prevent the marker genes to be removed during the graph pruning we can make the algorithm aware of them using the `marker_genes` parameter.

```{r biclustering_prune, eval = TRUE}

# Marker genes from A. Zeisel et al., Cell types in the mouse cortex and hippocampus revealed by single-cell RNA-seq, 2015, Science
markers <- c("Gad1",
             "Tbr1",
             "Spink8",
             "Mbp",
             "Aldoc",
             "Aif1",
             "Cldn5",
             "Acta2")

# SNN graph & biclustering
cabic <- caclust(obj = caobj,
                 k = c(70, 70, 70, 70),
                 select_genes = TRUE,
                 prune_overlap = TRUE,
                 mode = "all",
                 overlap = 0.2,
                 algorithm = 'leiden',
                 resolution = 1,
                 marker_genes = markers)
```
After building the  cell-gene graph, the biclustering is performed with either the Leiden algorithm (`algorithm = "leiden"`) or Spectral clustrering (`algorithm = "spectral"`).

The resulting caclust object contains biclustering results and the SNN graph:
```{r print_cabic}
cabic
```

The cluster assignments for the cells or genes can be obtained with the functions `gene_clusters()` and `cell_clusters()`:
```{r clusters}

sce$cabinet <- cell_clusters(cabic)

cat("Gene clusters:\n")
head(gene_clusters(cabic))

cat("\nCell clusters:\n")
head(cell_clusters(cabic))
```

If too many genes are included in the graph, CAbiNet may detect some clusters which only contain genes. These genes probably are not marker genes and do not help in annotating the cells. We can therefore safely remove these mono-clusters by running:
```{r rm_monoclusters}
cabic <- rm_monoclusters(cabic)
cabic
```
As in this example there weren't any mono-clusters, the results are identical to what we saw before.
Removing monoclusters should be done *before* computing the biMAP. As an added benefit, removing the mono-clusters often improves the layout of the embedding.

## Visualization of cells and genes in the biMAP

We now compute the biMAP embedding and visualize it in a scatter plot:
```{r bimap}

# Compute biMAP embedding.
cabic <- biMAP(cabic, k = 30)

# plot results
plot_biMAP(cabic,
           color_genes = TRUE,
           label_marker_genes = markers)

# only plot the cell clusters
plot_scatter_biMAP(cabic,
                   gene_alpha = 0.5,
                   color_genes = FALSE,
                   color_by = "level1class",
                   meta_df = colData(sce))
```

`plot_biMAP` and `plot_scatter_biMAP` both produce a scatter plot of the embedding. For very large data sets it can become difficult to differentiate between cells and genes in such a plot. If we are mainly interested in identifying the marker genes of the clusters, we can instead use the hex biMAP for a less busy visualization:

```{r hex_biMAP}
plot_hex_biMAP(cabic, color_genes = TRUE, hex_n = 20)
```


## Cell type annotation

CAbiNet allows for intuitive cell type annotation. Instead of having to make a large number of feature plots, we can simply display the position of a set of marker genes directly in the biMAP. By the position of genes and cells in the biMAP, we can easily tell the marker genes of each cell group without having to perform differential gene expression analysis.

```{r}

cts <- c("Interneurons",
         "S1 Pyramidal",
         "CA 1 Pyramidal",
         "Oligodendrocytes",
         "Astrocytes",
         "Microglia",
         "Endothelial",
         "Mural")

## make sure the cell types and marker genes are in a one-to-one order.
df <- data.frame(cell_type = cts,
                 marker_gene = markers)

genecls <- gene_clusters(cabic)
biclusters <- vector(mode="numeric", length = nrow(df))

for (i in seq_along(df$marker_gene)){
  biclusters[i] <- genecls[df$marker_gene[i]]
}

df$cabinet <- biclusters
df

```



```{r}
sce$cabinet_ct <- as.numeric(sce$cabinet)

for (i in seq_along(df$cabinet)){
  sel <- which(sce$cabinet_ct == df$cabinet[i])
  sce$cabinet_ct[sel] <- df$cell_type[i]
}
sce$cabinet_ct <- as.factor(sce$cabinet_ct)

plot_biMAP(cabic,
           meta_df = colData(sce),
           color_by = "cabinet_ct",
           label_marker_genes = markers)
```
The only remaining unnanotated cluster is bicluster 8. This cluster consists mostly of cells that have been annotated as astrocytes, but also oligodendrocytes and pyramidal CA1 cells. It could be that this is a either a distinct sub-cell type or a group of outliers/doublets.
We can also see the similar cell types Endothelial and Mural have been clustered in a single cluster (labelled Endothelial).

Using a feature biMAP to visualize the expression of the two Marker genes Acta2 (Mural) and Cldn5 (Endothelial) we can confirm that this cluster could be further subdivided by the location of the marker genes.

```{r biMAP_feature}
p1 <- plot_feature_biMAP(sce = sce,
                   caclust = cabic,
                   feature = 'Acta2',
                   label_size = 4)

p2 <- plot_feature_biMAP(sce = sce,
                   caclust = cabic,
                   feature = 'Cldn5',
                   label_size = 4)

p1
p2
```

If you do not have a list of marker genes for the cell types in your data, you can explore the visualization interactively, or just display the names of the marker genes in the plot. Co-clustered genes can be labeled in the biMAP. For example, if you don't have a list of known marker genes at hand, you can display the names of potential marker genes by
```{r label_mkg}
plot_biMAP(cabic, color_by = "cluster",
           label_marker_genes = TRUE,
           group_label_size = 6,
           color_genes = TRUE,
           max.overlaps = 20)
```

Sometimes the name of genes overlap with each other and can be difficult to read the names in the plot. Exploring the biMAP interactively allows you to zoom in or out at a group of points to inspect the information there in more detail:
```{r label_mkg_live}
plot_biMAP(cabic, color_by = "cluster",
           color_genes = TRUE,
           interactive = TRUE)
```

# Integration with Bioconductor packages

When intergrating the clustering results and cell type annotations with the SCE object, gene expression levels can also be visualized by violin plots.
```{r downstream}
scater::plotExpression(sce, features = markers[markers %in% rownames(sce)]  ,
                       x = 'cabinet', colour_by = 'cabinet') +
                       theme(axis.text.x = element_text(angle = 0, hjust = 1))
scater::plotExpression(sce, features = markers[markers %in% rownames(sce)]  ,
                       x = 'level1class', colour_by = 'level1class') +
                       theme(axis.text.x = element_text(angle = 30, hjust = 1))

```
Since CAbiNet only co-clusters the up-regulated genes with cell groups, the information of down-regulated genes is missing. To find out all the dysregulated genes, you can make use of the cell clustering results from CAbiNet and differential gene expression results from other packges to study the dysregulated genes for different cell clusters.


# Session info {.unnumbered}

```{r sessionInfo, echo=FALSE}
sessionInfo()
```
